src/Mod/TechDraw/App/ShapeExtractor.cpp

```diff
@@ -88,23 +91,67 @@ TopoDS_Shape ShapeExtractor::getShapes(const std::vector<App::DocumentObject*> l
             // Base::Console().Message("SE::getShapes - skipping 2d link: %s\n", l->Label.getValue());
             continue;
         }
-        if (l->isDerivedFrom<App::Link>()) {
-            App::Link* xLink = dynamic_cast<App::Link*>(l);
+
+        // Copy the pointer as not const so it can be changed if needed.
+        App::DocumentObject* obj = l;
+
+        bool isExplodedView = false;
+        auto proxy = dynamic_cast<App::PropertyPythonObject*>(obj->getPropertyByName("Proxy"));
+        if (proxy && proxy->getValue().hasAttr("saveAssemblyAndExplode")) {
+            isExplodedView = true;
+
+            Py::Object explodedViewPy = proxy->getValue();
+
+            Py::Object attr = explodedViewPy.getAttr("saveAssemblyAndExplode");
+            if (attr.ptr() && attr.isCallable()) {
+                Py::Tuple args(1);
+                args.setItem(0, Py::asObject(obj->getPyObject()));
```

Comment:
@wwmayer Excuse me but I have a crash that I cannot understand here. Perhaps you could help me understand please?
Doing `PyObject* test = obj->getPyObject()` is crashing freecad.

What I don't understand is that this does not always happen : 
- When I create a TechDraw view of the assembly-ExplodedView: it works OK.
- When I click on the 'Recompute' tool : it works OK.
- When I trigger TD 'Redraw Page' : It crashes.
- When I do 'Recompute Object' (so recompute only the TD view) : It crashes.
I cannot understand why this getPyObject sometimes works and sometime crashes. Any idea?

Reply:
The code looks correct to me. But you can check two things:
* is obj always a correct pointer or can it be dangling in some cases?
* the implementation of getPyObject() of the corresponding class could be wrong. By convention it must return a Python object with increased reference counter.

Reply:
Just before we there is  : 

```
        auto proxy = dynamic_cast<App::PropertyPythonObject*>(obj->getPropertyByName("Proxy"));
        if (proxy && proxy->getValue().hasAttr("saveAssemblyAndExplode")) {
            isExplodedView = true;

            Py::Object explodedViewPy = proxy->getValue();
            Py::Object attr = explodedViewPy.getAttr("saveAssemblyAndExplode");

            if (attr.ptr() && attr.isCallable()) {
```
So I would think the pointer is valid by that point, or can it be dangling?

The object is a FeaturePython created from python. So the function I guess would be : 
```
template<> PyObject* App::FeaturePython::getPyObject() {
    if (PythonObject.is(Py::_None())) {
        // ref counter is set to 1
        PythonObject = Py::Object(new FeaturePythonPyT<DocumentObjectPy>(this),true);
    }
    return Py::new_reference_to(PythonObject);
}
```

But I'm not sure to understand why it's using a template.

Reply:
> So I would think the pointer is valid by that point, or can it be dangling?

A dangling pointer causes undefined behaviour where everything can happen. Sometimes it passes and crashes later somewhere else. But when it always reaches the call of getPyObject() then it's rather unlikely that obj is a dangling pointer.

> But I'm not sure to understand why it's using a template.

This is needed to create the correct Python wrapper for a Python feature class.

Can you describe how to reproduce the crash?